### PR TITLE
Improve React Editor & Preview

### PR DESCRIPTION
**Pull request by djobbo**

### CODE REVIEW

File: app/app/globals.css [hunk 1]
```diff
@@ -59,3 +61,30 @@
...
+}
+
+.markdown li:not(:last-child) {
+  @apply pb-2;
+}
+
+.markdown code,
+code.inline {
+  @apply px-1 py-px bg-[#0D1117] rounded-sm text-sm text-[#ffffffdd];
+}
```
Comment (djobbo) [inline app/app/globals.css:90]: `markdown.css`?

File: app/app/layout.tsx [hunk 2]
```diff
@@ -23,7 +23,7 @@ export default async function RootLayout({
     <html lang="en" className="h-full">
       <body
         className={cn(
-          "h-full min-h-dvh bg-background font-sans antialiased",
+          "h-full min-h-dvh bg-[#161B22] font-sans antialiased",
```
Comment (djobbo) [inline app/app/layout.tsx:26]: don't hardcode colors

File: app/app/page.tsx [hunk 3]
```diff
@@ -1,50 +1,65 @@
...
 } from "@/components/ui/resizable"
-import { Preview } from "@/components/Preview"
-import templates from "@maruhq/templates"
+import Link from "next/link"
+import { Suspense } from "react"
+import Markdown from "react-markdown"
+import { challenges } from "@/features/challenges"
 
-export default async function Home() {
+export default async function EditorPage() {
```
Comment (djobbo) [inline app/app/page.tsx:13]: Home?

File: app/app/page.tsx [hunk 4]
```diff
@@ -1,50 +1,65 @@
...
-            <Editor
-              defaultCode={
-                template.src.directory["Component.tsx"].file.contents
-              }
-            />
+        <ResizablePanelGroup direction="horizontal" className="flex-1 p-2">
+          <ResizablePanel defaultSize={18} minSize={15} className="p-2">
+            <div className="markdown">
+              <Markdown>{challenge.description}</Markdown>
+            </div>
```
Comment (djobbo) [inline app/app/page.tsx:41]: create dedicated `<Markdown/>` component that wraps content with the `.markdown` class

File: app/features/editor/helpers/fetchExtraLibs.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+        fetch(url)
+          .then((res) => res.text())
+          .then((content) => [url, content] as const)
+          .catch(() => null),
+      ),
+    )
+
+    return libsContent.filter((lib): lib is ExtraLib => !!lib)
+  },
+)
```
Comment (djobbo) [inline app/features/editor/helpers/fetchExtraLibs.ts:23]: use raw-loader instead?

File: app/features/editor/helpers/saveEditorChanges.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
+import { type editor } from "monaco-editor/esm/vs/editor/editor.api"
+import { formatCode } from "./formatCode"
+
+export const saveEditorChanges = async (
+  editor: editor.IStandaloneCodeEditor,
+) => {
+  const cursorPosition = editor.getPosition()
+
+  const editorModel = editor.getModel()
```
Comment (djobbo) [inline app/features/editor/helpers/saveEditorChanges.ts:9]:
```suggestion
  const cursorPosition = editor.getPosition()
  const editorModel = editor.getModel()
```

File: app/features/editor/styles/editor.css [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  @apply border-l border-dashed border-l-[#ffffff11] !important;
+}
+
+.monaco-editor .current-line {
+  @apply rounded-sm !important;
+}
+
+.monaco-editor .monaco-hover {
+  @apply bg-card text-card-foreground border border-border rounded-lg shadow-md shadow-background p-1 !important;
+}
```
Comment (djobbo) [inline app/features/editor/styles/editor.css:39]: `!important` 😱

File: app/features/challenges/simple-counter/Component.tsx [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+          <svg
+            className="w-4 h-4 ml-2 fill-none stroke-current stroke-2 stroke"
+            xmlns="http://www.w3.org/2000/svg"
+            viewBox="0 0 24 24"
+            strokeLinecap="round"
+            strokeLinejoin="round"
+          >
+            <path d="M5 12h14" />
+            <path d="M12 5v14" />
+          </svg>
```
Comment (JoachimFarabet) [inline app/features/challenges/simple-counter/Component.tsx:24]: It's unclear what this SVG represents or its purpose within the button component. Maybe we want to place it in a separate component to provide better organization and clarity in the code.

File: app/features/editor/components/EditorContent.tsx [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+type EditorProps = {
+  extraLibs?: ExtraLib[]
+}
+
+export const EditorContent = ({ extraLibs = [] }: EditorProps) => {
+  const { code, setCode, renderPreview } = useEditor()
+  const editorRef = useRef<editor.IStandaloneCodeEditor | null>(null)
+
+  const handleCodeChange = (value: string = "") => {
```
Comment (JoachimFarabet) [inline app/features/editor/components/EditorContent.tsx:24]: we actually do not need "string"

File: app/features/editor/components/EditorContent.tsx [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+    extraLibs.forEach(([path, content]) => {
+      monaco.languages.typescript.typescriptDefaults.addExtraLib(content, path)
+    })
+  }
+
+  const handleKeyDown: KeyboardEventHandler = async (event) => {
+    const editor = editorRef.current
+    if (!editor) return
+
+    if (event.key === "s" && (event.ctrlKey || event.metaKey)) {
```
Comment (JoachimFarabet) [inline app/features/editor/components/EditorContent.tsx:60]:
```suggestion
    if ((event.ctrlKey || event.metaKey) && event.key === "s" ) {
```

File: app/app/layout.tsx (hunk 2 above)
Comment (djobbo) [reply app/app/layout.tsx:26]: Out of scope
resolved in #14

File: app/features/editor/styles/editor.css (hunk 7 above)
Comment (djobbo) [reply app/features/editor/styles/editor.css:39]: Out of scope
may be resolved in #14 ?

File: app/features/editor/helpers/fetchExtraLibs.ts (hunk 5 above)
Comment (djobbo) [reply app/features/editor/helpers/fetchExtraLibs.ts:23]: 🙅

**Pull request merged**